ouroboros-consensus/src/consensus-testlib/Test/Ouroboros/Consensus/ChainGenerator/Honest.hs

```diff
@@ -0,0 +1,647 @@
+
+{-
+
+Derivation of the transition matrix used in 'stationaryWindowSize'.
+The x/d summand is the chance we're sliding the window off of an active slot.
+The (d - x)/d summand is the chance we're sliding the window off of an empty slot.
+
+One reflecting barrier at n.
+
+    P( n -> n     ) = (n/d) * 1         +   ((d - n)/d) * (1 - f)
+    P( n -> n + 1 ) = (n/d) * 0         +   ((d - n)/d) * f
+
+And another at d.
+
+    P( d -> d - 1 ) = (d/d) * (1 - f)  +    (0/d) * f
+    P( d -> d     ) = (d/d) * f        +    (0/d) * (1 - f)
+
+Otherwise, stuttering random walk for n < x < d.
+
+    P( x -> x - 1 ) = (x/d) * (1 - f)   +   ((d - x)/d) * 0
+    P( x -> x     ) = (x/d) * f         +   ((d - x)/d) * (1 - f)
+    P( x -> x + 1 ) = (x/d) * 0         +   ((d - x)/d) * f
+
+Algebraically simplified.
+
+    d * P( n -> n     ) = d - df + nf
+    d * P( n -> n + 1 ) =     df - nf
+
+    d * P( x -> x - 1 ) = x - xf
+    d * P( x -> x     ) = 2xf + d - df - x
+    d * P( x -> x + 1 ) = df - xf
+
+        P( d -> d - 1 ) = 1 - f
+        P( d -> d     ) = f
+
+-}
+
+-----
+
+data EhcgViolation hon =
+    forall skolem.
+    EhcgViolation {
+        -- | How many active slots 'ehcgvWindow' has
+        ehcgvPopCount :: !(C.Size (C.Win EhcgLbl skolem) ActiveSlotE)
+      ,
+        -- | The ChainGrowth window that doesn't have enough active slots
+        ehcgvWindow   :: !(C.Contains SlotE hon (C.Win EhcgLbl skolem))
+      }
+
+instance Eq (EhcgViolation hon) where
+    EhcgViolation l1 l2 == EhcgViolation r1 r2 =
+        Some.runEq
+      $ Some.eqCtor EhcgViolation EhcgViolation
+            `Some.eqArg` (C.forgetBase, l1, C.forgetBase, r1)
+            `Some.eqArg` (C.forgetWindow, l2, C.forgetWindow, r2)
+
+instance Show (EhcgViolation hon) where
+    showsPrec p (EhcgViolation x y) =
+        Some.runShowsPrec p
+      $ Some.showCtor EhcgViolation "EhcgViolation"
+            `Some.showArg` x
+            `Some.showArg` y
+
+instance Read (EhcgViolation hon) where
+    readPrec =
+        Some.runReadPrec
+      $ Some.readCtor EhcgViolation "EhcgViolation"
+            <*> Some.readArg
+            <*> Some.readArg
+
+data HonestChainViolation hon =
+    -- | The chain must have at least one active slot
+    BadCount
+  |
+    BadEhcgWindow !(EhcgViolation hon)
+  |
+    BadLength !(C.Size hon SlotE)
+  deriving (Eq, Read, Show)
+
+-- | An Enriched Honest Chain Growth window
+data EhcgLbl
+
+-- | Check the Praos Enriched Chain Growth property
+--
+-- Definition of /window/ and /anchored window/. A window is a contiguous
+-- sequence of slots. A window anchored at a block starts with the slot
+-- immediately after that block.
+--
+-- Definition of /Praos Chain Growth Assumption/. We assume the honest chain
+-- contains at least @k@ blocks in every window that contains @s@ slots.
+--
+-- Definition of /Praos Enriched Chain Growth Assumption/. We assume the honest
+-- chain contains at least @k+1@ blocks in every window that contains @s - Δ@
```

Comment:
After discussing it with Facundo, I think the subtraction by delta in the denominator is likely unnecessary. It's only necessary if we assume the adversary can (significantly) accelerate immediately after the stability window.

The adversarial chain generator et al does allow for that to happen. But it's unlikely to be true on any real chain.

Reply:
"Praos Enriched Chain Growth Assumption" should be changed to "Length of Competing Chains" as in the test plan. The concern about subtracting or not subtracting $\Delta$ then becomes irrelevant, I think.

Reply:
Actually, the check of the adversarial chain was checking "Length of Competing Chains" although with a different name: "Praos Race Assumption".

In cd5aff46 I replaced the EHCG window with the stability window. The concerns about races are only checked for the schemas of the adversarial chains.